builtin/provisioners/puppet/resource_provisioner.go

```diff
@@ -0,0 +1,337 @@
+		case "winrm":
+			p.OSType = "windows"
+		default:
+			return fmt.Errorf("Unsupported connection type: %s", connType)
+		}
+	}
+
+	switch p.OSType {
+	case "nix":
+		p.runPuppetAgent = p.nixRunPuppetAgent
+		p.installPuppetAgent = p.nixInstallPuppetAgent
+		p.uploadFile = p.nixUploadFile
+		p.defaultCertname = p.nixDefaultCertname
+		p.isPuppetEnterprise = p.nixIsPuppetEnterprise
+	case "windows":
+		p.runPuppetAgent = p.windowsRunPuppetAgent
+		p.installPuppetAgent = p.windowsInstallPuppetAgent
+		p.uploadFile = p.windowsUploadFile
+		p.UseSudo = false
+		p.defaultCertname = p.windowsDefaultCertname
+		p.isPuppetEnterprise = p.windowsIsPuppetEnterprise
+	default:
+		return fmt.Errorf("Unsupported OS type: %s", p.OSType)
+	}
+
+	comm, err := communicator.New(state)
+	if err != nil {
+		return err
+	}
+
+	retryCtx, cancel := context.WithTimeout(ctx, comm.Timeout())
+	defer cancel()
+
+	err = communicator.Retry(retryCtx, func() error {
+		return comm.Connect(output)
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := comm.Disconnect(); err != nil {
+			rerr = err
+		}
+	}()
+
+	p.comm = comm
+
+	_, ok := configData.GetOkExists("open_source")
+	if !ok {
+		isPE, err := p.isPuppetEnterprise()
+		if err != nil {
+			return fmt.Errorf("Unable to determine if %s is running Puppet Enterprise: %v", p.Server, err)
+		}
+		p.OpenSource = !isPE
+	}
+
+	if p.OpenSource {
+		p.installPuppetAgent = p.installPuppetAgentOpenSource
+	}
+
+	csrAttrs := new(csrAttributes)
+	csrAttrs.CustomAttributes = make(map[string]string)
+	csrAttrs.ExtensionRequests = make(map[string]string)
+
+	if p.Autosign {
+		if p.Certname == "" {
+			p.Certname, _ = p.defaultCertname()
+		}
+
+		autosignToken, err := p.generateAutosignToken(p.Certname)
+		if err != nil {
+			return fmt.Errorf("Failed to generate an autosign token: %s", err)
+		}
+		csrAttrs.CustomAttributes["challengePassword"] = autosignToken
+	}
+
+	if p.PPRole != "" {
+		csrAttrs.ExtensionRequests["pp_role"] = p.PPRole
+	}
+
+	if err = p.writeCSRAttributes(csrAttrs); err != nil {
+		return fmt.Errorf("Failed to write csr_attributes.yaml: %s", err)
+	}
+
+	if err = p.installPuppetAgent(); err != nil {
+		return err
+	}
+
+	if err = p.runPuppetAgent(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateFn(config *terraform.ResourceConfig) (ws []string, es []error) {
+	return ws, es
+}
+
+func (p *provisioner) writeCSRAttributes(attrs *csrAttributes) (rerr error) {
+	file, err := ioutil.TempFile("", "puppet-csr-attrs")
+	if err != nil {
+		return fmt.Errorf("Failed to create a temp file: %s", err)
+	}
+	defer func() {
+		if err := os.Remove(file.Name()); err != nil {
+			rerr = err
+		}
+	}()
+
+	content, err := yaml.Marshal(attrs)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal CSR attributes to YAML: %s", err)
+	}
+
+	_, err = file.WriteString(string(content))
+	if err != nil {
+		return fmt.Errorf("Failed to write YAML to temp file: %s", err)
+	}
+
+	_, err = file.Seek(0, 0)
```

Comment:
I'm not very familiar with this pattern, can you tell me why you're going back to the start of the file here? It doesn't appear to be written to again.

Reply:
Ah ok, thank you!